### PR TITLE
use windows in 'time on page' queries

### PR DESCRIPTION
**Pull request by ruslandoga**
### Changes

This WIP PR makes 'time on page' queries use window functions instead of [neighbor.](https://clickhouse.com/docs/en/sql-reference/functions/other-functions#neighbor)

<details><summary>Why?</summary>

There is another WIP export/import PR https://github.com/ruslandoga/analytics/pull/160#discussion_r1366844625 that uses window functions and it seemed reasonable to try out this approach in dashboard queries as well.

</details>

**TODOs:**

- [x] keep prev approach, hide new approach behind a feature flag `:window_time_on_page`
- [ ] check performance on staging
- [ ] check performance on synthetic data
- [ ] remove `where p.pathname in ^pages` https://github.com/ruslandoga/analytics/pull/170
- [ ] check performance again

### Tests
- [x] Automated tests have been added

### Changelog
- [x] This PR does not make a user-facing change

### Documentation
- [x] This change does not need a documentation update

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

**Comment by ruslandoga** [comment]
https://github.com/plausible/analytics/blob/957138a8ec819380c17086f1594c2cdaab868119/test/plausible_web/controllers/api/stats_controller/pages_test.exs#L775-L834 fails. Can someone please explain me why the time on page is not 1600 (900 native + 700 imported)?

```elixir
  2) test GET /api/stats/:domain/pages calculates bounce rate and time on page for pages with imported data (PlausibleWeb.Api.StatsController.PagesTest)
     test/plausible_web/controllers/api/stats_controller/pages_test.exs:775
     Assertion with == failed
     code:  assert json_response(conn, 200) == [
              %{
                "bounce_rate" => 40.0,
                "time_on_page" => 800.0,
                "visitors" => 3,
                "pageviews" => 3,
                "name" => "/"
              },
              %{
                "bounce_rate" => nil,
                "time_on_page" => 60,
                "visitors" => 2,
                "pageviews" => 2,
                "name" => "/some-other-page"
              }
            ]
     left:  [
              %{"bounce_rate" => 40.0, "name" => "/", "pageviews" => 3, "time_on_page" => 1600.0, "visitors" => 3},
              %{"bounce_rate" => nil, "name" => "/some-other-page", "pageviews" => 2, "time_on_page" => 60.0, "visitors" => 2}
            ]
     right: [
              %{"bounce_rate" => 40.0, "name" => "/", "pageviews" => 3, "time_on_page" => 800.0, "visitors" => 3},
              %{"bounce_rate" => nil, "name" => "/some-other-page", "pageviews" => 2, "time_on_page" => 60, "visitors" => 2}
            ]
     stacktrace:
       test/plausible_web/controllers/api/stats_controller/pages_test.exs:818: (test)
```

File: test/plausible_web/controllers/api/stats_controller/pages_test.exs [hunk 1]
```diff
@@ -187,14 +187,14 @@ defmodule PlausibleWeb.Api.StatsController.PagesTest do
                  "visitors" => 2,
                  "pageviews" => 2,
                  "bounce_rate" => 0,
-                 "time_on_page" => 120.0
```
Comment (ruslandoga) [inline test/plausible_web/controllers/api/stats_controller/pages_test.exs:190]: I think it's incorrect, and should've been 240 seconds (1 minute + 3 minutes)

Comment (ruslandoga) [reply test/plausible_web/controllers/api/stats_controller/pages_test.exs:190]: Ah, it's not total, but avg. 🙈

File: lib/plausible/stats/aggregate.ex [hunk 2]
```diff
@@ -58,82 +58,58 @@ defmodule Plausible.Stats.Aggregate do
...
+               ),
+             timestamp: e.timestamp,
+             pathname: e.pathname
+           }
+
+    timed_pages_q =
+      from e in subquery(windowed_pages_q),
+        select: fragment("avgIf(?,?)", e.next_timestamp - e.timestamp, e.next_timestamp != 0)
+
+    timed_filtered_pages_q =
```
Comment (ukutaht) [inline lib/plausible/stats/aggregate.ex:83]: I *think* this can be replaced with:

```elixir
dynamic_where = Plausible.Stats.Base.dynamic_filter_condition(query, :page, "pathname")
from(timed_pages_q, where: ^dynamic_where)
```

but you'll first need to make the function from Base publicly accessible.

Comment (ruslandoga) [reply lib/plausible/stats/aggregate.ex]: Done in https://github.com/plausible/analytics/pull/3446/commits/2c58b873e7b66b8910c417d4bae47de96dc82e42

**Comment by ruslandoga** [comment]
@ukutaht I've moved the new approach behind a feature flag (which is enabled in the tests) and tried to minimize the changes. I think it can be now tried out on staging :)

**Comment by ukutaht** [comment]
Testing this locally I get completely different results for time_on_page depending on the flag:

Screenshot 2023-11-01 at 16 31 19

vs

Screenshot 2023-11-01 at 16 31 13

I haven't looked into why the difference might be. But it should be exactly the same logic, right?

**Comment by ruslandoga** [comment]
Yes, it should be exactly the same. I'll try testing manually tomorrow as well.

**Comment by ukutaht** [comment]
Means our test suite is insufficient as well...

**Comment by ruslandoga** [comment]
I've reproduced it on my instance. Now looking into it. I'll try to add a failing test which passes with `neighbor` and fails with windows.

**Comment by ruslandoga** [comment]
Some data for https://who.copycat.fun/stats.copycat.fun

<details><summary>Windows query calculates time_on_page as 589.5</summary>

```sql
SELECT
    avgIf(s0.next_timestamp-s0.timestamp,s0.next_timestamp!=0) as time_on_page
FROM (
    SELECT
        leadInFrame(se0.timestamp) OVER (
            PARTITION BY se0.session_id
            ORDER BY se0.timestamp
            ROWS BETWEEN CURRENT ROW AND 1 FOLLOWING
        ) AS next_timestamp,
        se0.timestamp,
        se0.pathname
    FROM "events_v2" AS se0
    sample 20000000
    WHERE (se0.site_id = 3)
        AND ((se0.timestamp >= 1693846800) AND (se0.timestamp < 1696525200))
) AS s0
WHERE s0.pathname = '/';

-- ┌─time_on_page─┐
-- │        589.5 │
-- └──────────────┘
```

</details>

<details><summary>Neighbors calculate time_on_page as NULL</summary>

```sql
SELECT
    avg(ifNotFinite(avgTime, null)) as time_on_age
FROM (
    SELECT
        p,
        sum(td)/count(case when p2 != p then 1 end) as avgTime
    FROM (
        SELECT
            p,
            p2,
            sum(t2-t) as td
        FROM (
            SELECT
                *,
                neighbor(t, 1) as t2,
                neighbor(p, 1) as p2,
                neighbor(s, 1) as s2
            FROM (
                SELECT
                    e0.pathname as p,
                    e0.timestamp as t,
                    e0.session_id as s
                FROM "events_v2" AS e0
                sample 20000000
                WHERE (e0.site_id = 3)
                    AND ((e0.timestamp >= 1693846800) AND (e0.timestamp < 1696525200))
                ORDER BY e0.session_id,e0.timestamp
            )
        )
        WHERE s=s2 AND p = '/'
        GROUP BY p,p2,s
    )
    GROUP BY p
);

-- ┌─time_on_age─┐
-- │        ᴺᵁᴸᴸ │
-- └─────────────┘
```

</details>

<details><summary>"Raw" data</summary>

```sql
SELECT
    groupArray((pathname, timestamp)),
    session_id
FROM events_v2
WHERE (site_id = 3) AND (timestamp >= 1693846800) AND (timestamp < 1696525200)
GROUP BY session_id
```
```plain
┌─groupArray(tuple(pathname, timestamp))────────────────────────────────────────────────┬───────────session_id─┐
│ [('/','2023-09-11 02:54:31')]                                                         │  4806447251051889137 │
│ [('/','2023-10-01 10:34:45')]                                                         │  7343302942351284309 │
│ [('/','2023-09-15 02:59:30'),('/','2023-09-15 02:59:47'),('/','2023-09-15 03:00:08')] │ 17382206299883481430 │
│ [('/','2023-09-07 04:24:10')]                                                         │ 11786982536106711858 │
│ [('/','2023-09-13 05:16:42')]                                                         │  5301652740193495217 │
│ [('/','2023-09-18 08:27:21')]                                                         │ 17868340381700394525 │
│ [('/','2023-09-09 07:31:38')]                                                         │  2013285148236388267 │
│ [('/','2023-09-09 08:23:11')]                                                         │  8898696346729502797 │
│ [('/','2023-10-03 06:47:20')]                                                         │ 14838826095873756608 │
│ [('/','2023-09-07 11:15:01')]                                                         │  6332333267146113377 │
│ [('/','2023-09-20 17:17:49')]                                                         │ 12911473700843647027 │
│ [('/','2023-09-17 15:59:02')]                                                         │ 18407311177393938561 │
│ [('/','2023-09-22 06:57:53')]                                                         │  5747400338488618351 │
│ [('/','2023-09-27 05:41:13')]                                                         │  5329007591594348031 │
│ [('/','2023-09-29 23:41:24')]                                                         │ 11737653659804444912 │
│ [('/','2023-09-20 09:51:07')]                                                         │ 14144008674124066109 │
│ [('/','2023-09-09 08:56:40')]                                                         │ 14690444992200609546 │
│ [('/','2023-09-11 12:31:31')]                                                         │   731233103301758184 │
│ [('/','2023-09-28 09:18:43')]                                                         │  4992453298077940663 │
│ [('/','2023-10-01 04:14:53')]                                                         │  2710776979681662931 │
│ [('/','2023-09-20 09:25:40'),('/','2023-09-20 09:29:06')]                             │ 11827054078857744174 │
│ [('/','2023-09-09 13:56:46')]                                                         │   866412391600667755 │
│ [('/','2023-09-22 10:45:14')]                                                         │  6180908294632689578 │
│ [('/','2023-09-30 16:21:38')]                                                         │ 12741399877130876260 │
│ [('/','2023-09-06 07:14:42'),('/','2023-09-06 07:39:08')]                             │ 11219203526942856378 │
│ [('/','2023-09-07 09:05:59')]                                                         │  3000544050302617491 │
│ [('/','2023-09-30 07:44:37')]                                                         │ 14057041236453112322 │
│ [('/','2023-09-06 16:21:45')]                                                         │  2029764983379646439 │
│ [('/','2023-09-11 06:42:05')]                                                         │  8647312894416775851 │
│ [('/','2023-09-23 09:17:48')]                                                         │   655559847236535964 │
│ [('/','2023-09-09 12:00:56')]                                                         │ 12370184673926199048 │
│ [('/','2023-10-01 14:19:15')]                                                         │ 16623035026953054930 │
│ [('/','2023-09-21 03:40:24')]                                                         │ 12774541621675132188 │
│ [('/','2023-09-07 08:27:00')]                                                         │  9744573245716483167 │
│ [('/','2023-09-19 11:02:59')]                                                         │ 13721837803397983040 │
│ [('/','2023-09-30 06:59:05')]                                                         │ 17288247430085649915 │
│ [('/','2023-09-13 13:54:26')]                                                         │  5818047431060821391 │
│ [('/','2023-09-13 12:38:48')]                                                         │  9692885827107712011 │
│ [('/','2023-09-22 02:14:07')]                                                         │  6473756118048615812 │
│ [('/','2023-09-10 11:38:56')]                                                         │  8415935364237608584 │
│ [('/','2023-09-13 09:01:13')]                                                         │ 13827694551596575565 │
│ [('/','2023-09-20 16:02:28')]                                                         │  3339046757890064283 │
│ [('/','2023-09-06 11:49:12')]                                                         │  3361098997515417208 │
│ [('/','2023-09-23 11:59:43')]                                                         │  2146182707432037495 │
│ [('/','2023-09-23 08:38:24')]                                                         │  6053505776193405420 │
│ [('/','2023-09-29 13:05:33'),('/','2023-09-29 13:20:29'),('/','2023-09-29 13:36:00')] │  3144267550722113218 │
│ [('/','2023-10-02 09:34:12')]                                                         │   835654674803658034 │
│ [('/','2023-09-19 09:20:36')]                                                         │ 13263212888496331458 │
│ [('/','2023-09-05 08:59:54')]                                                         │ 14190867948177347477 │
│ [('/','2023-09-20 10:57:11')]                                                         │ 13013299444385156245 │
│ [('/','2023-09-12 12:28:07')]                                                         │  8979666134341258852 │
│ [('/','2023-09-29 09:26:17')]                                                         │  8006246525438550495 │
│ [('/','2023-09-09 14:55:09')]                                                         │  1694237956921823538 │
│ [('/','2023-09-15 14:54:32')]                                                         │  1351214312956988676 │
│ [('/','2023-10-01 11:23:50')]                                                         │  2445506507262049058 │
└───────────────────────────────────────────────────────────────────────────────────────┴──────────────────────┘
```

</details>

**Comment by ruslandoga** [comment]
I think I messed up the avgs again 🙈

**Comment by ruslandoga** [comment]
I've added failing tests (failing on this branch, but passing on master) in https://github.com/plausible/analytics/pull/3494

Once https://github.com/ruslandoga/analytics/pull/183 is merged, they'd be passing in this branch as well.

**Comment by ruslandoga** [comment]
@ukutaht now that the failing tests https://github.com/plausible/analytics/pull/3494 and the fixes https://github.com/ruslandoga/analytics/pull/183 have been merged, I'm more confident that the logic is exactly the same. So much so that I'm re-requesting a review.

**Comment by ukutaht** [comment]
Locally I still see a difference:

On master:
Screenshot 2023-11-13 at 11 42 35

This PR:
Screenshot 2023-11-13 at 11 45 27

**Comment by ruslandoga** [comment]
Hm, can you please share the dataset? I checked on my instance's data and the results were the same...

**Comment by ruslandoga** [comment]
Oops, nevermind. If I select "all time" view, the results are different on my data as well.

**Comment by ukutaht** [comment]
Weird how many small inconsistencies there can be in this query... I know it's frustrating but at least this process will bring clarity and tests around all the little edge cases so we can be intentional about them going forward. Thanks for debugging these cases @ruslandoga !

File: test/plausible_web/controllers/api/stats_controller/pages_test.exs [hunk 3]
```diff
@@ -801,6 +801,39 @@ defmodule PlausibleWeb.Api.StatsController.PagesTest do
                |> json_response(200)
     end
 
+    test "calculates time on page per unique transition within session", %{conn: conn, site: site} do
```
Comment (ruslandoga) [inline test/plausible_web/controllers/api/stats_controller/pages_test.exs:804]: I've added another failing test and going to add the fix next.

Comment (ruslandoga) [reply test/plausible_web/controllers/api/stats_controller/pages_test.exs:804]: I don't think the name of this test is descriptive enough, but I'm not sure what else to call it.

File: test/plausible_web/controllers/api/stats_controller/pages_test.exs [hunk 4]
```diff
@@ -801,6 +801,39 @@ defmodule PlausibleWeb.Api.StatsController.PagesTest do
...
+      # ┌─p─────┬─p2──────┬─minus(t2, t)─┬──s─┐
+      # │ /a    │ /b      │          100 │ s1 │
+      # │ /a    │ /d      │          100 │ s2 │ <- this two get treated
+      # │ /a    │ /d      │            0 │ s2 │ <- as single page transition
+      # └───────┴─────────┴──────────────┴────┘
+      # so that 200 gets divided by 2 instead of 3
```
Comment (ruslandoga) [inline test/plausible_web/controllers/api/stats_controller/pages_test.exs:810]: So I've added a "pre-final-aggregation-table" here for illustration.

File: test/plausible_web/controllers/api/stats_controller/pages_test.exs (hunk 3 above)
Comment (ruslandoga) [reply test/plausible_web/controllers/api/stats_controller/pages_test.exs:804]: I've also added them as a separate PR into master: https://github.com/plausible/analytics/pull/3521

**Review by ukutaht — APPROVED** [review]
